File: Chapter 5/CSharpCarLibrary/Car.cs
using System;
using System.Text;
using System.Windows.Forms;

namespace CSharpCarLibrary
{
	/// <summary>
	/// Car color enum.
	/// </summary>
	public enum CarColor
	{
		Red, Green, Blue,
		Pink, Yellow, Black
	}

	/// <summary>
	/// Some classes in this library
	/// have a convertible...
	/// </summary>
	public interface IConvertible
	{
		void OpenSunRoof(bool openIt);
	}

	/// <summary>
	/// This is the base class for the autos
	/// in the library.
	/// </summary>
	public abstract class Car
	{
		// Polymorphic interface.
		public abstract void SpeedUp();
		public virtual void DisplayBumperStickerText()
		{
			MessageBox.Show("If you can read this you're too close.", "C# Car Library");
		}

		// State data.
		protected string mPetName;
		protected CarColor mCarColor;
		protected int mCurrSpeed;

		#region Class Constructors
		public Car(){}

		public Car(string name, CarColor color) 
			: this(name, color, 0){}

		public Car(string name, CarColor color, int sp)
		{
			mPetName = name;
			mCarColor = color;
			mCurrSpeed = sp;
		}
		#endregion

		#region Properties
		public string PetName
		{
			get{ return mPetName;}
			set{ mPetName = value;}
		}
		public CarColor Color
		{
			get{ return mCarColor;}
			set{ mCarColor = value;}
		}
		public int Speed
		{get{ return mCurrSpeed;}}
		#endregion

		// Change ToString() for Cars.
		public override string ToString()
		{
			StringBuilder sb = new StringBuilder();
			sb.AppendFormat("[C#] PetName: {0} Color: {1} CurrentSpeed: {2}", 
						    mPetName, mCarColor, mCurrSpeed);
			return sb.ToString();
		}
	}
}

File: Chapter 5/VbNetCarLibrary/Car.vb
' Car color.
Public Enum CarColor
    Red
    Green
    Blue
    Pink
    Yellow
    Black
End Enum

' Some cars are convertibles.
Public Interface IConvertible
    Sub OpenSunRoof(ByVal openIt As Boolean)
End Interface

' The base class.
Public MustInherit Class Car

    ' Polymorphic interface.
    Public MustOverride Sub SpeedUp()
    Public Overridable Sub DisplayBumperStickerText()
        MessageBox.Show("If you can read this you're too close.", "VB.NET Car Library")
    End Sub

    ' State data.
    Protected mPetName As String
    Protected mCarColor As CarColor
    Protected mCurrSpeed As Integer

#Region "Class Constructors"
    Public Sub New()
    End Sub

    Public Sub New(ByVal name As String, ByVal color As CarColor)
        MyClass.New(name, color, 0)
    End Sub

    Public Sub New(ByVal name As String, ByVal color As CarColor, ByVal sp As Integer)
        mPetName = name
        mCarColor = color
        mCurrSpeed = sp
    End Sub
#End Region

#Region "Properties"
    Public Property PetName() As String
        Get
            Return mPetName
        End Get
        Set(ByVal Value As String)
            mPetName = Value
        End Set
    End Property
    Public Property Color() As CarColor
        Get
            Return mCarColor
        End Get
        Set(ByVal Value As CarColor)
        End Set
    End Property
    Public Property Speed() As Integer
        Get
            Return mCurrSpeed
        End Get
        Set(ByVal Value As Integer)
            mCurrSpeed = Value
        End Set
    End Property
#End Region

    ' Change ToString() for Cars.
    Public Overrides Function ToString() As String
        Dim sb As StringBuilder = New StringBuilder()
        sb.AppendFormat("[VB.NET] PetName: {0} Color: {1} CurrentSpeed: {2}", _
          mPetName, mCarColor, mCurrSpeed)
        Return sb.ToString()
    End Function
End Class